### PR TITLE
Dashboard Schema v2: Transformation to v2 supports annotations

### PR DESCRIPTION
**Pull request by harisrozajac**
Added support for annotations when transforming scene to schemav2

This is a follow up on: https://github.com/grafana/grafana/pull/95546

### CODE REVIEW

File: public/app/features/dashboard-scene/serialization/__snapshots__/transformSceneToSaveModelSchemaV2.test.ts.snap [hunk 1]
```diff
@@ -2,7 +2,66 @@
 
 exports[`transformSceneToSaveModelSchemaV2 should transform scene to save model schema v2 1`] = `
 {
-  "annotations": [],
+  "annotations": [
+    {
+      "kind": "AnnotationQuery",
+      "spec": {
+        "datasource": {
```
Comment (harisrozajac) [inline public/app/features/dashboard-scene/serialization/__snapshots__/transformSceneToSaveModelSchemaV2.test.ts.snap:9]: datasource can be undefined because we are making it optional in veneer here (https://github.com/grafana/grafana/blob/20d2bd934376853a33c605e6b5ddb55af4fa914c/packages/grafana-schema/src/veneer/dashboard.types.ts#L37).

File: public/app/features/dashboard-scene/serialization/transformSceneToSaveModelSchemaV2.ts [hunk 2]
```diff
@@ -370,6 +380,36 @@ function getVariables(oldDash: DashboardSceneState) {
...
+  const annotations: AnnotationQueryKind[] = [];
+  for (const layer of data.state.annotationLayers) {
+    if (!(layer instanceof dataLayers.AnnotationsDataLayer)) {
+      continue;
+    }
+    const result: AnnotationQueryKind = {
+      kind: 'AnnotationQuery',
+      spec: {
+        name: layer.state.query.name,
+        datasource: layer.state.query.datasource!, // FIXME: this cannot be undefined
```
Comment (dprokop) [inline public/app/features/dashboard-scene/serialization/transformSceneToSaveModelSchemaV2.ts:397]: If no data source is provided on the query this means it will resolve to the default data source. I think we should loosen the type on the AnnotationQueryKind in order to let the runtime resolve the datasource as it does today.

Alternatively we could assume lack of the data source means "use default data source", but this will be problematic and make the dashboard json less portable. Suddenly the default data source becomes "fixed" in the save model.

Thoughts @ivanortegaalba or @torkelo?

File: public/app/features/dashboard-scene/serialization/transformSceneToSaveModelSchemaV2.ts [hunk 3]
```diff
@@ -370,6 +380,36 @@ function getVariables(oldDash: DashboardSceneState) {
...
+        query: {
+          kind: layer.state.query.datasource?.type ?? 'default',
```
Comment (dprokop) [inline public/app/features/dashboard-scene/serialization/transformSceneToSaveModelSchemaV2.ts:399]: So this needs cannot be 'default', it needs to be a plugin name. With the lack of the datasource provided we need to resolve the type of the default datasource.

File: public/app/features/dashboard-scene/serialization/transformSceneToSaveModelSchemaV2.ts [hunk 4]
```diff
@@ -370,6 +380,36 @@ function getVariables(oldDash: DashboardSceneState) {
...
+          spec: layer.state.query,
+        },
+        enable: Boolean(layer.state.isEnabled),
+        hide: Boolean(layer.state.isHidden),
+        filter: layer.state.query.filter ?? defaultAnnotationPanelFilter(),
+        iconColor: layer.state.query.iconColor,
```
Comment (dprokop) [inline public/app/features/dashboard-scene/serialization/transformSceneToSaveModelSchemaV2.ts:405]: We need an indiction here as well to say whether or not a annotation is built in. Currently we handle this on the query level (there's a property called `builtIn` on the default annotations provided by grafana). I think we need to lift it to AnnotationQueryKind spec.

File: public/app/features/dashboard-scene/serialization/__snapshots__/transformSceneToSaveModelSchemaV2.test.ts.snap [hunk 5]
```diff
@@ -2,7 +2,66 @@
...
+          "kind": "prometheus",
+          "spec": {
+            "datasource": {
+              "type": "prometheus",
+              "uid": "abcdef",
+            },
+            "enable": true,
+            "iconColor": "blue",
+            "name": "query2",
+          },
```
Comment (dprokop) [inline public/app/features/dashboard-scene/serialization/__snapshots__/transformSceneToSaveModelSchemaV2.test.ts.snap:60]: This doesn't feel right. See the spec proposal in the design doc: https://docs.google.com/document/d/1H9cnQJqdcxYFi2juEt5CBkiSRnCr4paqznDBYiHCKqM/edit?tab=t.0#heading=h.1bs4nm6bogp2. `query: DataQuery` should not have information about the data source, that information is provided by higher level spec - in this case it's the `AnnotationQuery`

File: public/app/features/dashboard-scene/serialization/transformSceneToSaveModelSchemaV2.ts (hunk 2 above)
Comment (ivanortegaalba) [reply public/app/features/dashboard-scene/serialization/transformSceneToSaveModelSchemaV2.ts:397]: Evaluating the two options, I see pros and cons here:

- Use a reserved keyword for default DS (`__default__`) enables differentiation between "no datasource" and "default" since they are two possible scenarios that can exist during runtime. 
- If we export with no DS annotation defined, we should remove that annotation query since this is not reproducible in another system. Also, it is not valid to save. So we could safely assume it exists or remove it. In this case, the default should be already the default ds ref, not a keyword.
- If we import an annotation without DS, we could assume it is the desired behavior by default, but this would be from the provisioning/as-code point of view.

So, in summary, I think `undefined` or `default` should only exist in "import/as-code" scenarios, where we want to fall back to default. Creating a keyword there forces users to learn there is a "keyword" to indicate the system to fall back to default, so I'd consider `undefined` as a possible value to import, but not an existent value to export. So I'd use undefined to indicate in a schema that the DS is the default one.

File: public/app/features/dashboard-scene/serialization/transformSceneToSaveModelSchemaV2.ts (hunk 3 above)
Comment (ivanortegaalba) [reply public/app/features/dashboard-scene/serialization/transformSceneToSaveModelSchemaV2.ts:399]: If we assume we want to handle `undefined` as default, yes. If we want to have a keyword, this `default` represents that keyword.

Comment (ivanortegaalba) [reply public/app/features/dashboard-scene/serialization/transformSceneToSaveModelSchemaV2.ts:399]: do you mean we need to resolve the data source during the transformation?

Comment (dprokop) [reply public/app/features/dashboard-scene/serialization/transformSceneToSaveModelSchemaV2.ts:399]: I think so - then you can retrieve the type of the data source.

Comment (ivanortegaalba) [reply public/app/features/dashboard-scene/serialization/transformSceneToSaveModelSchemaV2.ts:402]: This makes the whole process async, so it is not possible. I think we need to leave an empty object as a datasource type, and it is resolved in runtime once the scene is mounted. If DS service receives undefined or non-valid DS it falls back to default

File: packages/grafana-schema/src/schema/dashboard/v2alpha0/dashboard.gen.ts [hunk 6]
```diff
@@ -503,11 +503,9 @@ export const defaultVizConfigKind = (): VizConfigKind => ({
 });
 
 export interface AnnotationQuerySpec {
-	datasource: DataSourceRef;
+	datasource?: DataSourceRef;
 	query: DataQueryKind;
-	// TODO: Should be figured out based on datasource (Grafana ds)
-	// builtIn?: int
-	// Below are currently existing options for annotation queries
+	builtIn: boolean;
```
Comment (dprokop) [inline packages/grafana-schema/src/schema/dashboard/v2alpha0/dashboard.gen.ts:508]: Let's make it optional as it's only going to exist on one annotation query really.

File: public/app/features/dashboard-scene/serialization/transformSceneToSaveModelSchemaV2.ts (hunk 3 above)
Comment (dprokop) [reply public/app/features/dashboard-scene/serialization/transformSceneToSaveModelSchemaV2.ts:402]: Hm, but this means that we produce invalid kind and end up with incorrect schema. I guess the async thing makes it hard because we need to update many places to suddenly become async too?

Comment (ivanortegaalba) [reply public/app/features/dashboard-scene/serialization/transformSceneToSaveModelSchemaV2.ts:402]: yes. This translation should be synchronous, otherwise will make any process translating async, like creating a scene. The real thing is, that we have many async resolving for Scene, but once the scene is built and activated. So we should leave the scene to figure out what to put there when the kind is not defined. 

This is, in the end defining the type of the query. It is weird to not know it in advance, because even if the default is X, the type should be known in advance. So, falling back to default doesn't make much sense.

I don't really know how to solve this use case, tbh

Comment (dprokop) [reply public/app/features/dashboard-scene/serialization/transformSceneToSaveModelSchemaV2.ts:402]: @ivanortegaalba - first thing that comes to my mind is that we resolve the default datasource and keep in the dashboard scene state for retireval later on when serialising the state. Maybe we could keep it as part of the serialiser I'm proposing in https://github.com/grafana/grafana/pull/96931.

Alternatively we could allow providing a context to this function call, basically a 2nd argument, and then we would not have to make it async at all. However, whoever calls this function would have to be async.

Comment (dprokop) [reply public/app/features/dashboard-scene/serialization/transformSceneToSaveModelSchemaV2.ts:402]: >  It is weird to not know it in advance, because even if the default is X, the type should be known in advance

Well, technically, the default datasource is known at that time because it needs to be resolved for the queries to be run... It's just the function definition that makes it async...

But - the default data source can be resolved from boot config. There's the `defaultDatasource` uid in the config and the list of all the datasources from which we can figure out the default one synchroniously.

**Review by ivanortegaalba — APPROVED** [review]
I left a small comment. For the rest, LGTM!
